lib/DxilValidation/DxilContainerValidation.cpp

```diff
@@ -73,6 +78,610 @@ struct DiagRestore {
+  if (PSV1->ShaderStage == static_cast<uint8_t>(PSVShaderKind::Hull) ||
+      PSV1->ShaderStage == static_cast<uint8_t>(PSVShaderKind::Mesh)) {
+    uint32_t NumInputVectors = llvm::RoundUpToAlignment(NumInputScalars, 4) / 4;
+    uint32_t NumOutputVectors =
+        llvm::RoundUpToAlignment(NumPCOrPrimScalars, 4) / 4;
+    Mismatched |= ViewIDTableAndMaskMismatched(
+        PSV.GetInputToPCOutputTable(), ViewIDState.InputToPCOutputTable,
+        PSV.GetViewIDPCOutputMask(), ViewIDState.ViewIDPCOutputMask,
+        NumInputVectors, NumOutputVectors, PSV1->UsesViewID);
+  }
+
+  // NumPCOrPrimScalars as input, NumOutputScalars[0] as output.
+  if (PSV1->ShaderStage == static_cast<uint8_t>(PSVShaderKind::Domain)) {
+    uint32_t NumInputVectors =
+        llvm::RoundUpToAlignment(NumPCOrPrimScalars, 4) / 4;
+    uint32_t NumOutputVectors =
+        llvm::RoundUpToAlignment(NumOutputScalars[0], 4) / 4;
+    Mismatched |= ViewIDTableAndMaskMismatched(
+        PSV.GetPCInputToOutputTable(), ViewIDState.PCInputToOutputTable,
+        PSVComponentMask(), ViewIDState.ViewIDOutputMask[0], NumInputVectors,
+        NumOutputVectors, false);
+  }
+  if (Mismatched) {
+    std::string Str;
+    raw_string_ostream OS(Str);
+    PSV.PrintViewIDState(OS);
+    OS.flush();
+    std::string Str1;
+    raw_string_ostream OS1(Str1);
+    ViewIDState.Print(OS1, static_cast<PSVShaderKind>(PSV1->ShaderStage),
+                      PSV1->UsesViewID);
+    OS1.flush();
+    EmitMismatchError("ViewIDState", Str, Str1);
+  }
+}
+
+void PSVContentVerifier::VerifySignatures(unsigned ValMajor,
+                                          unsigned ValMinor) {
+  bool i1ToUnknownCompat = DXIL::CompareVersions(ValMajor, ValMinor, 1, 5) < 0;
+  // Verify input signature
+  VerifySignature(DM.GetInputSignature(), PSV.GetInputElement0(0),
+                  PSV.GetSigInputElements(), "SigInput", i1ToUnknownCompat);
+  // Verify output signature
+  VerifySignature(DM.GetOutputSignature(), PSV.GetOutputElement0(0),
+                  PSV.GetSigOutputElements(), "SigOutput", i1ToUnknownCompat);
+  // Verify patch constant signature
+  VerifySignature(DM.GetPatchConstOrPrimSignature(),
+                  PSV.GetPatchConstOrPrimElement0(0),
+                  PSV.GetSigPatchConstOrPrimElements(),
+                  "SigPatchConstantOrPrim", i1ToUnknownCompat);
+}
+
+void PSVContentVerifier::VerifySignature(const DxilSignature &Sig,
+                                         PSVSignatureElement0 *Base,
+                                         unsigned Count, std::string Name,
+                                         bool i1ToUnknownCompat) {
+  if (Count != Sig.GetElements().size()) {
+    EmitMismatchError(Name + "Elements", std::to_string(Count),
+                      std::to_string(Sig.GetElements().size()));
+    return;
+  }
+
+  ArrayRef<PSVSignatureElement0> Inputs(Base, Count);
+  // Build PSVSignatureSet with data in PSV.
+  PSVSignatureSet SigSet;
```

Comment:
This code allows signatures to differ in ordering between DxilModule and PSV0.  That should not be considered legal.  The ordering of signature elements must not differ.  What possible reason would you want to relax this?

Reply:
Fixed.